### ISSUE TITLE
no-restricted-imports: Allow only certain styles of imports (default / named / namespace)

### ISSUE THREAD

**Issue by kachkaev** [eslint/eslint#13874]
**What rule do you want to change?**

[no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports)

**Does this change cause the rule to produce more or fewer warnings?**

Same (the change is non-breaking)

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option: `importStyles: Array<"default" | "named" | "namespace">`

**Please provide some example code that this change will affect:**

```js
/*eslint no-restricted-imports: ["error", { paths: [{
    name: "react",
    importStyles: ["default", "named"],
    message: "Please use `import * as React from "react"` instead."
}]}]*/

// ❌
import React from "react";
import { useState } from "react";

// ✅
import * as React from "react";
```

**What does the rule currently do for this code?**

Currently, the rule does not help code collaborators to stick with only one specific style for imports for a given library. Thus, neighbouring files end up with a variety of import styles.

**What will the rule do after it's changed?**

Codebase owners will be able to limit imports styles for specific libraries. Thus, the ecosystem can get prepared to an upcoming removal of the default export in React.

For context, see: 
- https://twitter.com/dan_abramov/status/1308739731551858689
- https://epicreact.dev/importing-react-through-the-ages/

**Are you willing to submit a pull request to implement this change?**

Yep

**Comment by mdjermanovic** [comment]
Hi @kachkaev, thanks for the issue!

> Currently, the rule does not help code collaborators to stick with only one specific style for imports for a given library. Thus, neighbouring files end up with a variety of import styles.

I think this rule isn't supposed to enforce _style_, like the use of namespace imports over named imports.

There are plugin rules specifically for that purpose, e.g. [unicorn/import-style](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/import-style.md).

For the React use case, it can be configured as follows:

```js
/*eslint "unicorn/import-style": [
  "error",
  {
    "styles": {
      "react": {
        "namespace": true
      }
    }
  }
]*/

import React from "react"; // error
import { useState } from "react"; // error

import * as React_ from "react"; // ok
```

(added `_` just to avoid parsing error)

**Comment by kachkaev** [comment]
Thanks for the link to `"unicorn/import-style"` – I did not know this rule existed! At least there is a known plan B 👍

Speaking of adding `importStyle` to `"no-restricted-imports"`, which drawbacks do you see? At the moment, I [don’t use](https://github.com/kachkaev/routine-npm-packages/blob/2b0d29c2deadab48ec9d2eedee95e39d979927b8/packages/eslint-config-base/index.js) unicorn so adding it just for one rule would certainly create an overhead. It also seems reasonable to have the style of the import declared close to other restrictions. Having to search for an extra plugin adds up to the cost. If it was easy enough, I would not end up opening this issue.

Other folks and teams might be in a similar situation, so I wonder if the new `importStyles` option has merit in it 🤔 Happy to stick with `"unicorn/import-style"` if there are reasons to not add the new option.

**Comment by mdjermanovic** [comment]
> Speaking of adding `importStyle` to `"no-restricted-imports"`, which drawbacks do you see?

It looks to me more like a separate rule than an enhancement in `no-restricted-imports`, since it's about style rather than restricting the use of certain modules. Curious to hear opinions from other team members.

An alternative to plugin rules is the core [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax) rule, which can be configured for this case:

```js
/* eslint no-restricted-syntax: ["error",
  {
      selector: "ImportDeclaration[source.value='react'] > :matches(ImportSpecifier, ImportDefaultSpecifier)",
      message: "Please use `import * as React from 'react'` instead."
  }
]*/

// ❌
import React from "react";
import { useState } from "react";

// ✅
import * as React_ from "react";
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLFxuICB7XG4gICAgICBzZWxlY3RvcjogXCJJbXBvcnREZWNsYXJhdGlvbltzb3VyY2UudmFsdWU9J3JlYWN0J10gPiA6bWF0Y2hlcyhJbXBvcnRTcGVjaWZpZXIsIEltcG9ydERlZmF1bHRTcGVjaWZpZXIpXCIsXG4gICAgICBtZXNzYWdlOiBcIlBsZWFzZSB1c2UgYGltcG9ydCAqIGFzIFJlYWN0IGZyb20gJ3JlYWN0J2AgaW5zdGVhZC5cIlxuICB9XG5dKi9cblxuLy8g4p2MXG5pbXBvcnQgUmVhY3QgZnJvbSBcInJlYWN0XCI7XG5pbXBvcnQgeyB1c2VTdGF0ZSB9IGZyb20gXCJyZWFjdFwiO1xuXG4vLyDinIVcbmltcG9ydCAqIGFzIFJlYWN0XyBmcm9tIFwicmVhY3RcIjsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Comment by nzakas** [comment]
I agree with @mdjermanovic — I don’t think it makes sense to add style checks to the rule. This seems like a good use case for a custom rule.

**Comment by btmills** [comment]
I agree with @mdjermanovic and @nzakas that this option isn't a fit for this particular rule. Based on the Twitter thread you helpfully linked, I get the impression TypeScript already handles this for those who can use it. For plain JS, I'm glad to see `unicorn/import-style` works. If it's important to encourage widespread adoption of this check, the best vector is probably a dedicated rule in `eslint-plugin-react/recommended` to get it to as many developers as possible without them having to see a message about it and configure it for each project.

**Comment by nzakas** [comment]
Closing, as we won't be moving forward with this issue.

**Comment by kachkaev** [comment]
Thanks for your comments folks, the resolution makes sense. I did not know about [unicorn/import-style](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/import-style.md) when creating this issue and it is indeed a good enough solution 👍
